app/assets/javascripts/components/materials_bin/_fetch_data_mixin.js.coffee

```diff
@@ -0,0 +1,39 @@
+# Fetches remote data when component is visible (when @props.visible == true).
+# Data is saved under @state[@dataStateKey].
+# Client class has to define:
+#  - .dataUrl property (string)
+#  - .dataStateKey property (string), name of the state key under which data is saved
+# but it can also define:
+#  - .requestParams property (hash), argument of jQuery.ajax
+#  - .processData() (method), it can process raw AJAX response before state is updated
+window.MBFetchDataMixin =
+  getInitialState: ->
```

Comment:
Not a bug but using getIntialState in a mixin means it can't be mixed in with a component that also declares getInitialState as only one getInitialState declaration is allowed.  Your components are fine but if you need to do this in the future the way I've done it is to declare something like getInitialDataMixinState: (initialState) in the mixin and then merge the two states and return it.  Then in the component class getInitialState calls getInitialDataMixinState with the initial state as a parameter.  That way you can also chain multiple mixin states.